"unexpected EOF" causes Google Cloud workflows to fail

## Motivation

Sometimes Google Cloud CI workflows fail with:
> unexpected EOF
> Error: Process completed with exit code 1.

https://github.com/ZcashFoundation/zebra/runs/5728525849?check_suite_focus=true

This seems to be happening a few times a day. It might be a Google Cloud change or infrastructure bug.

## Suggested Fix

If the `ssh` command fails, wait a few seconds, then re-launch the command. Fail if the `ssh` command has failed more than 5-10 times.

This needs to be fixed in every workflow that follows Google Cloud logs.

--- a/.github/workflows/test-full-sync.yml
+++ b/.github/workflows/test-full-sync.yml
@@ -236,12 +236,15 @@ jobs:
       - name: Full sync
         id: full-sync
         run: |
-          gcloud compute ssh \
-          full-sync-${{ env.GITHUB_REF_SLUG_URL }}-${{ env.GITHUB_SHA_SHORT }} \
-          --zone ${{ env.ZONE }} \
-          --quiet \
-          --ssh-flag="-o ServerAliveInterval=5" \
-          --command="docker logs --follow ${{ env.CONTAINER_NAME }}"
+          for RETRY in 1 2 3 4; do
+              gcloud compute ssh \
+              full-sync-${{ env.GITHUB_REF_SLUG_URL }}-${{ env.GITHUB_SHA_SHORT }} \
+              --zone ${{ env.ZONE }} \
+              --quiet \
+              --ssh-flag="-o ServerAliveInterval=15" \
+              --command="docker logs --follow ${{ env.CONTAINER_NAME }}" \
+              || echo "ssh disconnected $RETRY times"
+          done
 
           EXIT_CODE=$(\
           gcloud compute ssh \

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -347,12 +347,15 @@ jobs:
         id: sync-to-checkpoint
         if: ${{ steps.create-instance.outcome == 'success' }}
         run: |
-          gcloud compute ssh \
-          regenerate-disk-${{ env.GITHUB_REF_SLUG_URL }}-${{ env.GITHUB_SHA_SHORT }} \
-          --zone ${{ env.ZONE }} \
-          --quiet \
-          --ssh-flag="-o ServerAliveInterval=5" \
-          --command="docker logs --follow ${{ env.CONTAINER_NAME }}"
+          for RETRY in 1 2 3 4; do
+              gcloud compute ssh \
+              regenerate-disk-${{ env.GITHUB_REF_SLUG_URL }}-${{ env.GITHUB_SHA_SHORT }} \
+              --zone ${{ env.ZONE }} \
+              --quiet \
+              --ssh-flag="-o ServerAliveInterval=15" \
+              --command="docker logs --follow ${{ env.CONTAINER_NAME }}" \
+              || echo "ssh disconnected $RETRY times"
+          done
 
           EXIT_CODE=$(\
           gcloud compute ssh \
@@ -512,12 +515,15 @@ jobs:
       - name: Sync past mandatory checkpoint
         id: sync-past-checkpoint
         run: |
-          gcloud compute ssh \
-          sync-checkpoint-${{ env.GITHUB_REF_SLUG_URL }}-${{ env.GITHUB_SHA_SHORT }} \
-          --zone ${{ env.ZONE }} \
-          --quiet \
-          --ssh-flag="-o ServerAliveInterval=5" \
-          --command="docker logs --follow ${{ env.CONTAINER_NAME }}"
+          for RETRY in 1 2 3 4; do
+              gcloud compute ssh \
+              sync-checkpoint-${{ env.GITHUB_REF_SLUG_URL }}-${{ env.GITHUB_SHA_SHORT }} \
+              --zone ${{ env.ZONE }} \
+              --quiet \
+              --ssh-flag="-o ServerAliveInterval=15" \
+              --command="docker logs --follow ${{ env.CONTAINER_NAME }}" \
+              || echo "ssh disconnected $RETRY times"
+          done
 
           EXIT_CODE=$(\
           gcloud compute ssh \

